src/concurrency/thread.rs

```diff
@@ -147,6 +153,28 @@ impl<'mir, 'tcx> Thread<'mir, 'tcx> {
     fn thread_name(&self) -> &[u8] {
         if let Some(ref thread_name) = self.thread_name { thread_name } else { b"<unnamed>" }
     }
+
+    fn get_top_user_relevant_frame(&self) -> Option<usize> {
+        self.stack
+            .iter()
+            .enumerate()
+            .rev()
+            .find_map(|(idx, frame)| if frame.extra.is_user_relevant { Some(idx) } else { None })
+    }
+
+    pub fn recompute_top_user_relevant_frame(&mut self) {
+        self.top_user_relevant_frame = self.get_top_user_relevant_frame();
+    }
+
+    pub fn set_top_user_relevant_frame(&mut self, frame_idx: usize) {
+        debug_assert_eq!(Some(frame_idx), self.get_top_user_relevant_frame());
+        self.top_user_relevant_frame = Some(frame_idx);
+    }
+
+    pub fn top_user_relevant_frame(&self) -> usize {
+        debug_assert_eq!(self.top_user_relevant_frame, self.get_top_user_relevant_frame());
+        self.top_user_relevant_frame.unwrap_or_else(|| self.stack.len().saturating_sub(1))
```

Comment:
Any reason not to include the `.unwrap_or_else(|| self.stack.len().saturating_sub(1))` in `get_top_user_relevant_frame`?

Ah, then it wouldn't be enough to just update this when we push a user-relevant frame, right? Seems subtle enough to be worth documenting.

Can this even be called when the stack is empty? Looks like `current_span` would just panic then anyway.

Reply:
> Ah, then it wouldn't be enough to just update this when we push a user-relevant frame, right? Seems subtle enough to be worth documenting.

Correct. Will do.

> Can this even be called when the stack is empty? Looks like current_span would just panic then anyway.

Yes this is called when the stack is empty. We query the current span when creating an `AllocHistory`, and we create allocations before the program has started. I'll try to document that.